### PR TITLE
[SPARK-37915][SQL] Combine unions if there is a project between them

### PR DESCRIPTION
**Pull request by wangyum**
### What changes were proposed in this pull request?

This pr makes `CombineUnions` combine unions if there is a project between them. For example:
```scala
spark.range(1).selectExpr("CAST(id AS decimal(18, 1)) AS id").write.saveAsTable("t1")
spark.range(2).selectExpr("CAST(id AS decimal(18, 2)) AS id").write.saveAsTable("t2")
spark.range(3).selectExpr("CAST(id AS decimal(18, 3)) AS id").write.saveAsTable("t3")
spark.range(4).selectExpr("CAST(id AS decimal(18, 4)) AS id").write.saveAsTable("t4")
spark.range(5).selectExpr("CAST(id AS decimal(18, 5)) AS id").write.saveAsTable("t5")

spark.sql("SELECT id FROM t1 UNION SELECT id FROM t2 UNION SELECT id FROM t3 UNION SELECT id FROM t4 UNION SELECT id FROM t5").explain(true)
```

Before this pr:
```
== Optimized Logical Plan ==
Aggregate [id#36], [id#36]
+- Union false, false
   :- Aggregate [id#34], [cast(id#34 as decimal(22,5)) AS id#36]
   :  +- Union false, false
   :     :- Aggregate [id#32], [cast(id#32 as decimal(21,4)) AS id#34]
   :     :  +- Union false, false
   :     :     :- Aggregate [id#30], [cast(id#30 as decimal(20,3)) AS id#32]
   :     :     :  +- Union false, false
   :     :     :     :- Project [cast(id#25 as decimal(19,2)) AS id#30]
   :     :     :     :  +- Relation default.t1[id#25] parquet
   :     :     :     +- Project [cast(id#26 as decimal(19,2)) AS id#31]
   :     :     :        +- Relation default.t2[id#26] parquet
   :     :     +- Project [cast(id#27 as decimal(20,3)) AS id#33]
   :     :        +- Relation default.t3[id#27] parquet
   :     +- Project [cast(id#28 as decimal(21,4)) AS id#35]
   :        +- Relation default.t4[id#28] parquet
   +- Project [cast(id#29 as decimal(22,5)) AS id#37]
      +- Relation default.t5[id#29] parquet
```

After this pr:
```
== Optimized Logical Plan ==
Aggregate [id#36], [id#36]
+- Union false, false
   :- Project [cast(id#25 as decimal(22,5)) AS id#36]
   :  +- Relation default.t1[id#25] parquet
   :- Project [cast(id#26 as decimal(22,5)) AS id#46]
   :  +- Relation default.t2[id#26] parquet
   :- Project [cast(id#27 as decimal(22,5)) AS id#45]
   :  +- Relation default.t3[id#27] parquet
   :- Project [cast(id#28 as decimal(22,5)) AS id#44]
   :  +- Relation default.t4[id#28] parquet
   +- Project [cast(id#29 as decimal(22,5)) AS id#37]
      +- Relation default.t5[id#29] parquet
```

### Why are the changes needed?

Improve query performance by reduce shuffles.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Unit test.

### CODE REVIEW

**Comment by wangyum** [comment]
cc @cloud-fan @viirya

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -766,18 +767,24 @@ object PushProjectionThroughUnion extends Rule[LogicalPlan] with PredicateHelper
...
-        p
+      val newFirstChild = Project(projectList, u.children.head)
+      val newOtherChildren = u.children.tail.map { child =>
+        val rewrites = buildRewrites(u.children.head, child)
+        Project(projectList.map(pushToRight(_, rewrites)), child)
+      }
+      u.copy(children = newFirstChild +: newOtherChildren)
+
+    // Push down deterministic projection through SQL UNION
+    case Project(projectList, Distinct(u: Union)) if projectList.forall(_.deterministic) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:780]: hmm, how does this help eliminate aggregates?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -78,7 +78,6 @@ abstract class Optimizer(catalogManager: CatalogManager)
     val operatorOptimizationRuleSet =
       Seq(
         // Operator push down
-        PushProjectionThroughUnion,
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:81]: This looks like a weird fix to me, by moving an optimizer rule from the main batch to a new batch before `Batch("Union", ...)`.

Can we fix `CombineUnions` instead? To support group-only `Aggregate`.

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:81]: +1

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 3]
```diff
@@ -762,22 +762,22 @@ object PushProjectionThroughUnion extends Rule[LogicalPlan] with PredicateHelper
...
+      Project(projectList.map(pushToRight(_, rewrites)), child)
+    }
+    newFirstChild +: newOtherChildren
+  }
+
   def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
     _.containsAllPatterns(UNION, PROJECT)) {
 
     // Push down deterministic projection through UNION ALL
-    case p @ Project(projectList, u: Union) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:769]: unnecessary change?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 4]
```diff
@@ -1322,6 +1322,12 @@ object CombineUnions extends Rule[LogicalPlan] {
         case Union(children, byName, allowMissingCol)
             if byName == topByName && allowMissingCol == topAllowMissingCol =>
           stack.pushAll(children.reverse)
+        case Project(projectList, Distinct(u @ Union(children, byName, allowMissingCol)))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1325]: It makes sense to me to accept Project between Unions when combining them. IIRC we did similar things for window optimization.

Shall we make it more general? It seems there are 3 forms of union: Union, Deduplicate + Union, Distinct + Union. Should all these three accept Projects in between?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 3 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:769]: `CombineUnions` need it to push down projection
https://github.com/apache/spark/pull/35214/files#diff-11264d807efa58054cca2d220aae8fba644ee0f0f2a4722c46d52828394846efR1328

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 5]
```diff
@@ -1322,6 +1322,12 @@ object CombineUnions extends Rule[LogicalPlan] {
...
+            if projectList.forall(_.deterministic) && children.nonEmpty &&
+              flattenDistinct && byName == topByName && allowMissingCol == topAllowMissingCol =>
+          val newChildren = PushProjectionThroughUnion.pushProjectionThroughUnion(projectList, u)
+            .map(CollapseProject(_))
```
Comment (wangyum) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1329]: `CollapseProject` is needed because there may be two `Project`s:
```
Project(projectList1, Project(projectList2, Distinct(u @ Union(children, byName, allowMissingCol))))
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1325]: It already supports `UNION ALL` but needs to be executed more times. We can use same logic to make it only execute once.
```
13:51:52.309 ERROR org.apache.spark.sql.catalyst.rules.PlanChangeLogger: 
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.CombineUnions ===
 Union false, false                                                                         Union false, false
!:- Union false, false                                                                      :- Project [cast(id#34 as decimal(22,5)) AS id#36]
!:  :- Project [cast(id#34 as decimal(22,5)) AS id#36]                                      :  +- Union false, false
!:  :  +- Union false, false                                                                :     :- Project [cast(id#32 as decimal(21,4)) AS id#34]
!:  :     :- Project [cast(id#32 as decimal(21,4)) AS id#34]                                :     :  +- Union false, false
!:  :     :  +- Union false, false                                                          :     :     :- Project [cast(cast(id#25 as decimal(19,2)) as decimal(20,3)) AS id#32]
!:  :     :     :- Project [cast(cast(id#25 as decimal(19,2)) as decimal(20,3)) AS id#32]   :     :     :  +- Relation default.t1[id#25] parquet
!:  :     :     :  +- Relation default.t1[id#25] parquet                                    :     :     +- Project [cast(cast(id#26 as decimal(19,2)) as decimal(20,3)) AS id#46]
!:  :     :     +- Project [cast(cast(id#26 as decimal(19,2)) as decimal(20,3)) AS id#46]   :     :        +- Relation default.t2[id#26] parquet
!:  :     :        +- Relation default.t2[id#26] parquet                                    :     +- Project [cast(cast(id#27 as decimal(20,3)) as decimal(21,4)) AS id#45]
!:  :     +- Project [cast(cast(id#27 as decimal(20,3)) as decimal(21,4)) AS id#45]         :        +- Relation default.t3[id#27] parquet
!:  :        +- Relation default.t3[id#27] parquet                                          :- Project [cast(cast(id#28 as decimal(21,4)) as decimal(22,5)) AS id#44]
!:  +- Project [cast(cast(id#28 as decimal(21,4)) as decimal(22,5)) AS id#44]               :  +- Relation default.t4[id#28] parquet
!:     +- Relation default.t4[id#28] parquet                                                +- Project [cast(id#29 as decimal(22,5)) AS id#37]
!+- Project [cast(id#29 as decimal(22,5)) AS id#37]                                            +- Relation default.t5[id#29] parquet
!   +- Relation default.t5[id#29] parquet                                                   
           
13:51:52.319 ERROR org.apache.spark.sql.catalyst.rules.PlanChangeLogger: 
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.SimplifyCasts ===
 Union false, false                                                                      Union false, false
 :- Project [cast(id#34 as decimal(22,5)) AS id#36]                                      :- Project [cast(id#34 as decimal(22,5)) AS id#36]
 :  +- Union false, false                                                                :  +- Union false, false
 :     :- Project [cast(id#32 as decimal(21,4)) AS id#34]                                :     :- Project [cast(id#32 as decimal(21,4)) AS id#34]
 :     :  +- Union false, false                                                          :     :  +- Union false, false
!:     :     :- Project [cast(cast(id#25 as decimal(19,2)) as decimal(20,3)) AS id#32]   :     :     :- Project [cast(id#25 as decimal(20,3)) AS id#32]
 :     :     :  +- Relation default.t1[id#25] parquet                                    :     :     :  +- Relation default.t1[id#25] parquet
!:     :     +- Project [cast(cast(id#26 as decimal(19,2)) as decimal(20,3)) AS id#46]   :     :     +- Project [cast(id#26 as decimal(20,3)) AS id#46]
 :     :        +- Relation default.t2[id#26] parquet                                    :     :        +- Relation default.t2[id#26] parquet
!:     +- Project [cast(cast(id#27 as decimal(20,3)) as decimal(21,4)) AS id#45]         :     +- Project [cast(id#27 as decimal(21,4)) AS id#45]
 :        +- Relation default.t3[id#27] parquet                                          :        +- Relation default.t3[id#27] parquet
!:- Project [cast(cast(id#28 as decimal(21,4)) as decimal(22,5)) AS id#44]               :- Project [cast(id#28 as decimal(22,5)) AS id#44]
 :  +- Relation default.t4[id#28] parquet                                                :  +- Relation default.t4[id#28] parquet
 +- Project [cast(id#29 as decimal(22,5)) AS id#37]                                      +- Project [cast(id#29 as decimal(22,5)) AS id#37]
    +- Relation default.t5[id#29] parquet                                                   +- Relation default.t5[id#29] parquet
           
13:51:52.328 ERROR org.apache.spark.sql.catalyst.rules.PlanChangeLogger: 
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.PushProjectionThroughUnion ===
 Union false, false                                               Union false, false
!:- Project [cast(id#34 as decimal(22,5)) AS id#36]               :- Union false, false
!:  +- Union false, false                                         :  :- Project [cast(id#34 as decimal(22,5)) AS id#36]
!:     :- Project [cast(id#32 as decimal(21,4)) AS id#34]         :  :  +- Union false, false
!:     :  +- Union false, false                                   :  :     :- Project [cast(id#32 as decimal(21,4)) AS id#34]
!:     :     :- Project [cast(id#25 as decimal(20,3)) AS id#32]   :  :     :  +- Project [cast(id#25 as decimal(20,3)) AS id#32]
!:     :     :  +- Relation default.t1[id#25] parquet             :  :     :     +- Relation default.t1[id#25] parquet
!:     :     +- Project [cast(id#26 as decimal(20,3)) AS id#46]   :  :     +- Project [cast(id#46 as decimal(21,4)) AS id#48]
!:     :        +- Relation default.t2[id#26] parquet             :  :        +- Project [cast(id#26 as decimal(20,3)) AS id#46]
!:     +- Project [cast(id#27 as decimal(21,4)) AS id#45]         :  :           +- Relation default.t2[id#26] parquet
!:        +- Relation default.t3[id#27] parquet                   :  +- Project [cast(id#45 as decimal(22,5)) AS id#47]
!:- Project [cast(id#28 as decimal(22,5)) AS id#44]               :     +- Project [cast(id#27 as decimal(21,4)) AS id#45]
!:  +- Relation default.t4[id#28] parquet                         :        +- Relation default.t3[id#27] parquet
!+- Project [cast(id#29 as decimal(22,5)) AS id#37]               :- Project [cast(id#28 as decimal(22,5)) AS id#44]
!   +- Relation default.t5[id#29] parquet                         :  +- Relation default.t4[id#28] parquet
!                                                                 +- Project [cast(id#29 as decimal(22,5)) AS id#37]
!                                                                    +- Relation default.t5[id#29] parquet
           
13:51:52.341 ERROR org.apache.spark.sql.catalyst.rules.PlanChangeLogger: 
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.CollapseProject ===
 Union false, false                                               Union false, false
 :- Union false, false                                            :- Union false, false
 :  :- Project [cast(id#34 as decimal(22,5)) AS id#36]            :  :- Project [cast(id#34 as decimal(22,5)) AS id#36]
 :  :  +- Union false, false                                      :  :  +- Union false, false
!:  :     :- Project [cast(id#32 as decimal(21,4)) AS id#34]      :  :     :- Project [cast(cast(id#25 as decimal(20,3)) as decimal(21,4)) AS id#34]
!:  :     :  +- Project [cast(id#25 as decimal(20,3)) AS id#32]   :  :     :  +- Relation default.t1[id#25] parquet
!:  :     :     +- Relation default.t1[id#25] parquet             :  :     +- Project [cast(cast(id#26 as decimal(20,3)) as decimal(21,4)) AS id#48]
!:  :     +- Project [cast(id#46 as decimal(21,4)) AS id#48]      :  :        +- Relation default.t2[id#26] parquet
!:  :        +- Project [cast(id#26 as decimal(20,3)) AS id#46]   :  +- Project [cast(cast(id#27 as decimal(21,4)) as decimal(22,5)) AS id#47]
!:  :           +- Relation default.t2[id#26] parquet             :     +- Relation default.t3[id#27] parquet
!:  +- Project [cast(id#45 as decimal(22,5)) AS id#47]            :- Project [cast(id#28 as decimal(22,5)) AS id#44]
!:     +- Project [cast(id#27 as decimal(21,4)) AS id#45]         :  +- Relation default.t4[id#28] parquet
!:        +- Relation default.t3[id#27] parquet                   +- Project [cast(id#29 as decimal(22,5)) AS id#37]
!:- Project [cast(id#28 as decimal(22,5)) AS id#44]                  +- Relation default.t5[id#29] parquet
!:  +- Relation default.t4[id#28] parquet                         
!+- Project [cast(id#29 as decimal(22,5)) AS id#37]               
!   +- Relation default.t5[id#29] parquet                         
           
13:51:52.348 ERROR org.apache.spark.sql.catalyst.rules.PlanChangeLogger: 
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.CombineUnions ===
 Union false, false                                                                   Union false, false
!:- Union false, false                                                                :- Project [cast(id#34 as decimal(22,5)) AS id#36]
!:  :- Project [cast(id#34 as decimal(22,5)) AS id#36]                                :  +- Union false, false
!:  :  +- Union false, false                                                          :     :- Project [cast(cast(id#25 as decimal(20,3)) as decimal(21,4)) AS id#34]
!:  :     :- Project [cast(cast(id#25 as decimal(20,3)) as decimal(21,4)) AS id#34]   :     :  +- Relation default.t1[id#25] parquet
!:  :     :  +- Relation default.t1[id#25] parquet                                    :     +- Project [cast(cast(id#26 as decimal(20,3)) as decimal(21,4)) AS id#48]
!:  :     +- Project [cast(cast(id#26 as decimal(20,3)) as decimal(21,4)) AS id#48]   :        +- Relation default.t2[id#26] parquet
!:  :        +- Relation default.t2[id#26] parquet                                    :- Project [cast(cast(id#27 as decimal(21,4)) as decimal(22,5)) AS id#47]
!:  +- Project [cast(cast(id#27 as decimal(21,4)) as decimal(22,5)) AS id#47]         :  +- Relation default.t3[id#27] parquet
!:     +- Relation default.t3[id#27] parquet                                          :- Project [cast(id#28 as decimal(22,5)) AS id#44]
!:- Project [cast(id#28 as decimal(22,5)) AS id#44]                                   :  +- Relation default.t4[id#28] parquet
!:  +- Relation default.t4[id#28] parquet                                             +- Project [cast(id#29 as decimal(22,5)) AS id#37]
!+- Project [cast(id#29 as decimal(22,5)) AS id#37]                                      +- Relation default.t5[id#29] parquet
!   +- Relation default.t5[id#29] parquet                                             
           
13:51:52.354 ERROR org.apache.spark.sql.catalyst.rules.PlanChangeLogger: 
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.SimplifyCasts ===
 Union false, false                                                                Union false, false
 :- Project [cast(id#34 as decimal(22,5)) AS id#36]                                :- Project [cast(id#34 as decimal(22,5)) AS id#36]
 :  +- Union false, false                                                          :  +- Union false, false
!:     :- Project [cast(cast(id#25 as decimal(20,3)) as decimal(21,4)) AS id#34]   :     :- Project [cast(id#25 as decimal(21,4)) AS id#34]
 :     :  +- Relation default.t1[id#25] parquet                                    :     :  +- Relation default.t1[id#25] parquet
!:     +- Project [cast(cast(id#26 as decimal(20,3)) as decimal(21,4)) AS id#48]   :     +- Project [cast(id#26 as decimal(21,4)) AS id#48]
 :        +- Relation default.t2[id#26] parquet                                    :        +- Relation default.t2[id#26] parquet
!:- Project [cast(cast(id#27 as decimal(21,4)) as decimal(22,5)) AS id#47]         :- Project [cast(id#27 as decimal(22,5)) AS id#47]
 :  +- Relation default.t3[id#27] parquet                                          :  +- Relation default.t3[id#27] parquet
 :- Project [cast(id#28 as decimal(22,5)) AS id#44]                                :- Project [cast(id#28 as decimal(22,5)) AS id#44]
 :  +- Relation default.t4[id#28] parquet                                          :  +- Relation default.t4[id#28] parquet
 +- Project [cast(id#29 as decimal(22,5)) AS id#37]                                +- Project [cast(id#29 as decimal(22,5)) AS id#37]
    +- Relation default.t5[id#29] parquet                                             +- Relation default.t5[id#29] parquet
           
13:51:52.360 ERROR org.apache.spark.sql.catalyst.rules.PlanChangeLogger: 
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.PushProjectionThroughUnion ===
 Union false, false                                         Union false, false
!:- Project [cast(id#34 as decimal(22,5)) AS id#36]         :- Union false, false
!:  +- Union false, false                                   :  :- Project [cast(id#34 as decimal(22,5)) AS id#36]
!:     :- Project [cast(id#25 as decimal(21,4)) AS id#34]   :  :  +- Project [cast(id#25 as decimal(21,4)) AS id#34]
!:     :  +- Relation default.t1[id#25] parquet             :  :     +- Relation default.t1[id#25] parquet
!:     +- Project [cast(id#26 as decimal(21,4)) AS id#48]   :  +- Project [cast(id#48 as decimal(22,5)) AS id#49]
!:        +- Relation default.t2[id#26] parquet             :     +- Project [cast(id#26 as decimal(21,4)) AS id#48]
!:- Project [cast(id#27 as decimal(22,5)) AS id#47]         :        +- Relation default.t2[id#26] parquet
!:  +- Relation default.t3[id#27] parquet                   :- Project [cast(id#27 as decimal(22,5)) AS id#47]
!:- Project [cast(id#28 as decimal(22,5)) AS id#44]         :  +- Relation default.t3[id#27] parquet
!:  +- Relation default.t4[id#28] parquet                   :- Project [cast(id#28 as decimal(22,5)) AS id#44]
!+- Project [cast(id#29 as decimal(22,5)) AS id#37]         :  +- Relation default.t4[id#28] parquet
!   +- Relation default.t5[id#29] parquet                   +- Project [cast(id#29 as decimal(22,5)) AS id#37]
!                                                              +- Relation default.t5[id#29] parquet
           
13:51:52.378 ERROR org.apache.spark.sql.catalyst.rules.PlanChangeLogger: 
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.CollapseProject ===
 Union false, false                                         Union false, false
 :- Union false, false                                      :- Union false, false
!:  :- Project [cast(id#34 as decimal(22,5)) AS id#36]      :  :- Project [cast(cast(id#25 as decimal(21,4)) as decimal(22,5)) AS id#36]
!:  :  +- Project [cast(id#25 as decimal(21,4)) AS id#34]   :  :  +- Relation default.t1[id#25] parquet
!:  :     +- Relation default.t1[id#25] parquet             :  +- Project [cast(cast(id#26 as decimal(21,4)) as decimal(22,5)) AS id#49]
!:  +- Project [cast(id#48 as decimal(22,5)) AS id#49]      :     +- Relation default.t2[id#26] parquet
!:     +- Project [cast(id#26 as decimal(21,4)) AS id#48]   :- Project [cast(id#27 as decimal(22,5)) AS id#47]
!:        +- Relation default.t2[id#26] parquet             :  +- Relation default.t3[id#27] parquet
!:- Project [cast(id#27 as decimal(22,5)) AS id#47]         :- Project [cast(id#28 as decimal(22,5)) AS id#44]
!:  +- Relation default.t3[id#27] parquet                   :  +- Relation default.t4[id#28] parquet
!:- Project [cast(id#28 as decimal(22,5)) AS id#44]         +- Project [cast(id#29 as decimal(22,5)) AS id#37]
!:  +- Relation default.t4[id#28] parquet                      +- Relation default.t5[id#29] parquet
!+- Project [cast(id#29 as decimal(22,5)) AS id#37]         
!   +- Relation default.t5[id#29] parquet                   
           
13:51:52.391 ERROR org.apache.spark.sql.catalyst.rules.PlanChangeLogger: 
=== Applying Rule org.apache.spark.sql.catalyst.optimizer.CombineUnions ===
 Union false, false                                                             Union false, false
!:- Union false, false                                                          :- Project [cast(cast(id#25 as decimal(21,4)) as decimal(22,5)) AS id#36]
!:  :- Project [cast(cast(id#25 as decimal(21,4)) as decimal(22,5)) AS id#36]   :  +- Relation default.t1[id#25] parquet
!:  :  +- Relation default.t1[id#25] parquet                                    :- Project [cast(cast(id#26 as decimal(21,4)) as decimal(22,5)) AS id#49]
!:  +- Project [cast(cast(id#26 as decimal(21,4)) as decimal(22,5)) AS id#49]   :  +- Relation default.t2[id#26] parquet
!:     +- Relation default.t2[id#26] parquet                                    :- Project [cast(id#27 as decimal(22,5)) AS id#47]
!:- Project [cast(id#27 as decimal(22,5)) AS id#47]                             :  +- Relation default.t3[id#27] parquet
!:  +- Relation default.t3[id#27] parquet                                       :- Project [cast(id#28 as decimal(22,5)) AS id#44]
!:- Project [cast(id#28 as decimal(22,5)) AS id#44]                             :  +- Relation default.t4[id#28] parquet
!:  +- Relation default.t4[id#28] parquet                                       +- Project [cast(id#29 as decimal(22,5)) AS id#37]
!+- Project [cast(id#29 as decimal(22,5)) AS id#37]                                +- Relation default.t5[id#29] parquet
!   +- Relation default.t5[id#29] parquet                                       

```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 5 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1330]: Next optimization iteration should run `CollapseProject` on it, no?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 4 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1326]: Hm? Above two optimized plans look different.

**Review by viirya — COMMENTED** [review]
Previously cannot PushProjectionThroughUnion + CombineUnions achieve same effect that combines unions with project between them? i.e.,

Original plan:
```
Project
  - Union
    - Union
      - Child1 
    - Union
      - Child2
    - Union
      - Child3 
    - ...
```

After `PushProjectionThroughUnion`:
```
Union
  - Project
    - Union
      - Child1 
  - Project
    - Union
      - Child2
  - Project
    - Union
      - Child3 
  - ...
```

Next iteration. After `PushProjectionThroughUnion`:
```
Union
  - Union
    - Project
      - Child1 
  - Union
    - Project
      - Child2
  - Union
    - Project
      - Child3 
  - ...
```

After `CombineUnions`:
```
Union
  - Project
    - Child1 
  - Project
    - Child2
  - Project
    - Child3 
  - ...
```

**Comment by wangyum** [comment]
Another attempt combine unions after `ReplaceDistinctWithAggregate`: https://github.com/apache/spark/pull/35249

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 5 above)
Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1330]: We will put all children to stack and use it later. For example, if we do not collapse projects. we can not [extract union‘s children](https://github.com/apache/spark/blob/ab18cc8a7d3df476f045f84bae4ff121eacc91b9/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L1326-L1343):
```
Project [cast(id#34 as decimal(22,5)) AS id#36]
+- Project [cast(id#32 as decimal(21,4)) AS id#34]
   +- Distinct
      +- Union false, false
         :- Project [cast(id#30 as decimal(20,3)) AS id#32]
         :  +- Distinct
         :     +- Union false, false
         :        :- Project [cast(id#25 as decimal(19,2)) AS id#30]
         :        :  +- Relation default.t1[id#25] parquet
         :        +- Project [cast(id#26 as decimal(19,2)) AS id#31]
         :           +- Relation default.t2[id#26] parquet
         +- Project [cast(id#27 as decimal(20,3)) AS id#33]
            +- Relation default.t3[id#27] parquet
```
And later we will do `ReplaceDistinctWithAggregate` and then `CombineUnions` can not combine any unions.

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1330]: Another attempt combine unions after `ReplaceDistinctWithAggregate`: https://github.com/apache/spark/pull/35249

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 6]
```diff
@@ -1322,6 +1322,25 @@ object CombineUnions extends Rule[LogicalPlan] {
...
+            if projectList.forall(_.deterministic) && flattenDistinct && u.byName == topByName &&
+              u.allowMissingCol == topAllowMissingCol && AttributeSet(keys) == u.outputSet =>
+          val newChildren = PushProjectionThroughUnion.pushProjectionThroughUnion(projectList, u)
+            .map(CollapseProject(_))
+          stack.pushAll(newChildren.reverse)
+        case Project(projectList, u @ Union(children, byName, allowMissingCol))
+            if projectList.forall(_.deterministic) && children.nonEmpty &&
+              byName == topByName && allowMissingCol == topAllowMissingCol =>
+          val newChildren = PushProjectionThroughUnion.pushProjectionThroughUnion(projectList, u)
+            .map(CollapseProject(_))
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1342]: I think it's a bit overkill to run the entire `CollapseProject` rule. It also makes it harder to understand the algorithm here. It's actually a simple algorithm: When we hit `Project Chain(Union(A, B))`, we put `Project Chain(A)` and `Project Chain(B)` back to the stack and repeat this process. In the end, we will union a bunch of plans with a very deep `Project` chain. To be more efficient, we should collapse projects during this process.

The code can be
```
stack.pop() match {
  case p1 @ Project(_, p2: Project) if CollapseProject.canCollapseExpressions(p1, p2) =>
    stack.push(p2.copy(projectList = CollapseProject.buildCleanedProjectList(p1.projectList, p2.projectList)))
  case Union(children, ...) => stack.pushAll(children.reverse)
  case Project(list, Union(children, ...)) if list.forall(_.deterministic) =>
    stack.pushAll(children.reverse.map(Project(list, _)))
  ... // handle distinct and deduplicate
}
```

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: Make sense.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 7]
```diff
@@ -1300,16 +1300,22 @@ object InferFiltersFromConstraints extends Rule[LogicalPlan]
...
-    case u: Union => flattenUnion(u, false)
-    case Distinct(u: Union) => Distinct(flattenUnion(u, true))
-    // Only handle distinct-like 'Deduplicate', where the keys == output
-    case Deduplicate(keys: Seq[Attribute], u: Union) if AttributeSet(keys) == u.outputSet =>
-      Deduplicate(keys, flattenUnion(u, true))
+  import CollapseProject.{buildCleanedProjectList, canCollapseExpressions}
+  import PushProjectionThroughUnion.pushProjectionThroughUnion
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    val alwaysInline = conf.getConf(SQLConf.COLLAPSE_PROJECT_ALWAYS_INLINE)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1307]: This is a new optimization. I don't think we need to respect this config for keeping legacy behaviors.

Comment (wangyum) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:1307]: +1

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Review by viirya — COMMENTED** [review]
lgtm too

**Comment by tanelk** [comment]
Hi,

I found a regression in spark 3.3.0 compared to spark 3.2.0. Git bisect lead me to this PR. It looks like this PR is not the direct cause, but it instead revealed an existing bug.

Minimal testcase:
```
  test("SPARK-XXXXX: Cache with join, union and renames") {
    val df1 = Seq("1", "2").toDF("a")
    val df2 = Seq("2", "3").toDF("a")
      .withColumn("b", lit("b"))

    val joined = df1.join(broadcast(df2), "a")
      // Messing around the column can cause some problems with cache manager
      .withColumn("tmp_b", $"b")
      .drop("b")
      .withColumnRenamed("tmp_b", "b")
      .cache()

    val unioned = joined.union(joined)

    assertCached(unioned, 2)
  }
```
Error:
```
[info]   ArrayBuffer() had size 0 instead of expected size 2 Expected query to contain 2, but it actually had 0
[info]   Union false, false
[info]   :- Project [a#4, b#12 AS b#23]
[info]   :  +- Join Inner, (a#4 = a#10)
[info]   :     :- Project [value#1 AS a#4]
[info]   :     :  +- LocalRelation [value#1]
[info]   :     +- ResolvedHint (strategy=broadcast)
[info]   :        +- Project [a#10, b AS b#12]
[info]   :           +- Project [value#7 AS a#10]
[info]   :              +- LocalRelation [value#7]
[info]   +- Project [a#4 AS a#38, b#23 AS b#39]
[info]      +- Project [a#4, b#12 AS b#23]
[info]         +- Join Inner, (a#4 = a#10)
[info]            :- Project [value#36 AS a#4]
[info]            :  +- LocalRelation [value#36]
[info]            +- ResolvedHint (strategy=broadcast)
[info]               +- Project [a#10, b AS b#12]
[info]                  +- Project [value#37 AS a#10]
[info]                     +- LocalRelation [value#37] (QueryTest.scala:198)
```

As you see, the `InMemoryRelation` is missing, but I have no idea where to start looking for it.

**Comment by wangyum** [comment]
This is a known issue, please see comment:
https://github.com/apache/spark/blob/69f402ad8085bc50837128645b61168e7d5244b9/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L2297-L2299

**Comment by cloud-fan** [comment]
`df.union` can break caching by design, but it was to optimize a special df pattern `df1.union(df2).union(df3).union...`. This PR does make it worse as `df.union` can break caching more. I'm working on a fix to make `df.union` targeting the special df pattern it was designed to optimize.
